src/python/pants/backend/helm/util_rules/plugins.py

```diff
@@ -0,0 +1,171 @@
+    description: str | None = None
+    ignore_flags: bool | None = None
+    command: str | None = None
+    platform_command: tuple[HelmPluginPlatformCommand, ...] = field(default_factory=tuple)
+    hooks: FrozenDict[str, str] = field(default_factory=FrozenDict)
+
+    @classmethod
+    def from_dict(cls, d: dict[str, Any]) -> HelmPluginMetadata:
+        platform_command = [
+            HelmPluginPlatformCommand.from_dict(d) for d in d.pop("platformCommand", [])
+        ]
+        hooks = d.pop("hooks", {})
+
+        attrs = yaml_attr_dict(d)
+        return cls(platform_command=tuple(platform_command), hooks=FrozenDict(hooks), **attrs)
+
+    @classmethod
+    def from_bytes(cls, content: bytes) -> HelmPluginMetadata:
+        return HelmPluginMetadata.from_dict(yaml.safe_load(content))
+
+
+_ExternalHelmPlugin = TypeVar("_ExternalHelmPlugin", bound=ExternalHelmPlugin)
+_GHP = TypeVar("_GHP", bound="ExternalHelmPluginBinding")
+
+
+@union
+@dataclass(frozen=True)
+class ExternalHelmPluginBinding(Generic[_ExternalHelmPlugin], metaclass=ABCMeta):
+    plugin_subsystem_cls: ClassVar[Type[ExternalHelmPlugin]]
+
+    name: str
+
+    @final
+    @classmethod
+    def create(cls: Type[_GHP]) -> _GHP:
+        return cls(name=cls.plugin_subsystem_cls.plugin_name)
+
+
+@dataclass(frozen=True)
+class ExternalHelmPluginRequest:
+    plugin_name: str
+    tool_request: ExternalToolRequest
+
+    @classmethod
+    def from_subsystem(cls, subsystem: ExternalHelmPlugin) -> ExternalHelmPluginRequest:
+        return cls(
+            plugin_name=subsystem.plugin_name, tool_request=subsystem.get_request(Platform.current)
+        )
+
+
+@dataclass(frozen=True)
+class HelmPlugin:
+    metadata: HelmPluginMetadata
+    digest: Digest
+
+    @property
+    def name(self) -> str:
+        return self.metadata.name
+
+    @property
+    def version(self) -> str:
+        return self.metadata.version
+
+
+class HelmPlugins(Collection[HelmPlugin]):
+    pass
+
+
+@rule
+async def all_helm_plugins(union_membership: UnionMembership) -> HelmPlugins:
+    bindings = union_membership.get(ExternalHelmPluginBinding)
+    external_plugins = await MultiGet(
+        Get(HelmPlugin, ExternalHelmPluginBinding, binding.create()) for binding in bindings
+    )
+    if logger.isEnabledFor(LogLevel.DEBUG.level):
+        plugins_desc = [f"{p.name}, version: {p.version}" for p in external_plugins]
+        logger.debug(
+            f"Downloaded {pluralize(len(external_plugins), 'external Helm plugin')}:\n{bullet_list(plugins_desc)}"
+        )
+    return HelmPlugins(external_plugins)
+
+
+@rule(desc="Download an external Helm plugin", level=LogLevel.DEBUG)
+async def download_external_helm_plugin(request: ExternalHelmPluginRequest) -> HelmPlugin:
+    downloaded_tool = await Get(DownloadedExternalTool, ExternalToolRequest, request.tool_request)
+
+    metadata_file = await Get(
+        Digest, DigestSubset(downloaded_tool.digest, PathGlobs(["plugin.yaml"]))
```

Comment:
While this code later in this function does check whether the file was found or not, consider using the `glob_match_error_behavior` parameter to `PathGlobs`.